be/src/olap/rowset/segment_v2/binary_prefix_page.cpp

```diff
@@ -0,0 +1,250 @@
+    return Status::OK();
+}
+
+OwnedSlice BinaryPrefixPageBuilder::finish() {
+    DCHECK(!_finished);
+    _finished = true;
+    int old_size = _buffer.size();
+    put_fixed32_le(&_buffer, (uint32_t)_count);
+    old_size += sizeof(uint32_t);
+    int restart_point_size = _restart_points_offset.size();
+    for(unsigned int i = 0; i < restart_point_size; ++i) {
+        put_fixed32_le(&_buffer, _restart_points_offset[i]);
+        old_size += sizeof(uint32_t);
+    }
+    put_fixed32_le(&_buffer, restart_point_size);
+    return _buffer.build();
+}
+
+const uint8_t* BinaryPrefixPageDecoder::_decode_value_lengths(const uint8_t* ptr,
+                                                              uint32_t* shared,
+                                                              uint32_t* non_shared) {
+    const uint8_t* limit = _restarts_ptr - sizeof(uint32_t);
+    if ((ptr = decode_varint32_ptr(ptr, limit, shared)) == nullptr) {
+        return nullptr;
+    }
+    if ((ptr = decode_varint32_ptr(ptr, limit, non_shared)) == nullptr) {
+        return nullptr;
+    }
+    if (limit - ptr < *non_shared) {
+        return nullptr;
+    }
+    return ptr;
+}
+
+Status BinaryPrefixPageDecoder::_read_next_value() {
+    if (_cur_pos >= _num_values) {
+        return Status::NotFound("no more value to read");
+    }
+    uint32_t shared_len;
+    uint32_t non_shared_len;
+    auto data_ptr = _decode_value_lengths(_next_ptr, &shared_len, &non_shared_len);
+    if (data_ptr == nullptr) {
+        return Status::Corruption(Substitute("Failed to decode value at position $0", _cur_pos));
+    }
+    _current_value.resize(shared_len);
+    _current_value.append(data_ptr, non_shared_len);
+    _next_ptr = data_ptr + non_shared_len;
+    return Status::OK();
+}
+
+Status BinaryPrefixPageDecoder::_seek_to_restart_point(size_t restart_point_index) {
+    _cur_pos = restart_point_index * RESTART_POINT_INTERVAL;
+    _next_ptr = _get_restart_point(restart_point_index);
+    return _read_next_value();
+}
+
+Status BinaryPrefixPageDecoder::init() {
+    _cur_pos = 0;
+    _next_ptr = reinterpret_cast<const uint8_t*>(_data.get_data());
+
+    const uint8_t* end = _next_ptr + _data.get_size();
+    _num_restarts = decode_fixed32_le(end - sizeof(uint32_t));
+    _restarts_ptr = end - (_num_restarts + 1) * sizeof(uint32_t);
+    _num_values = decode_fixed32_le(end - (_num_restarts + 2) * sizeof(uint32_t));
+    _parsed = true;
+    return _read_next_value();
+}
+
+Status BinaryPrefixPageDecoder::seek_to_position_in_page(size_t pos) {
+    DCHECK(_parsed);
+    DCHECK_LE(pos, _num_values);
+
+    // seek past the last value is valid
+    if (pos == _num_values) {
+        _cur_pos = _num_values;
+        return Status::OK();
+    }
+
+    size_t restart_point_index = pos / RESTART_POINT_INTERVAL;
+    RETURN_IF_ERROR(_seek_to_restart_point(restart_point_index));
+    while (_cur_pos < pos) {
+        _cur_pos++;
+        RETURN_IF_ERROR(_read_next_value());
```

Comment:
Do we need to copy unshared part for seek operation?

Reply:
Yes. We need. because the `share_len` and `non_share_len` will be different for every value, and we need to get `_current_value` after seek.